### PR TITLE
Use new C++ document selection parser

### PR DESCRIPTION
**Pull request by vekterli**
@baldersheim here comes a big one to review! This tries as much as possible to reproduce 1-1 the behavior of the legacy C++ parser, including its quirks. This causes some aspects of the parser to be.. less elegant than they theoretically could be.
@toregge @havardpe @arnej27959 FYI, I would greatly appreciate more sets of eyes on this.

Based on Flex/Bison and replaces old Spirit.Classic parser.
New parser is pure and does not require any locking, unlike the
previous implementation.

This also removes parsing of the deprecated searchcolumn feature.

Adds build-time dependencies on Flex and Bison.

### CODE REVIEW

**Comment by vekterli** [comment]
Hold off merging this until a newer version of Bison is included as part of our internal build dependencies (needs 3.0+). Flex version looks fine.

**Comment by toregge** [comment]
LGTM

**Review by baldersheim — APPROVED** [review]
👍

**Comment by vekterli** [comment]
As a FYI, the reason this PR remains unmerged is due to wanting the changes in #4186 to be verified as regression-free. I explicitly do _not_ want a release with changes to document selection across two different languages at the same time.

File: dist/vespa.spec [hunk 1]
```diff
@@ -68,6 +68,8 @@ BuildRequires: make
 BuildRequires: vespa-cppunit-devel >= 1.12.1-6
 BuildRequires: vespa-libtorrent-devel >= 1.0.11-6
 BuildRequires: systemd
+BuildRequires: flex >= 2.5.0
+BuildRequires: bison >= 3.0.0
```
Comment (aressem) [inline dist/vespa.spec]: Both of these requirements are ok on CentOS 7

**Comment by havardpe** [comment]
CMake seems to have issues finding Bison for pull request builds. Could this be related?

-- Found JNI: /usr/lib/jvm/java-openjdk/jre/lib/amd64/libjawt.so
CMake Error at /usr/share/cmake3/Modules/FindPackageHandleStandardArgs.cmake:148 (message):
Could NOT find BISON (missing: BISON_EXECUTABLE)
Call Stack (most recent call first):
/usr/share/cmake3/Modules/FindPackageHandleStandardArgs.cmake:388 (_FPHSA_FAILURE_MESSAGE)
/usr/share/cmake3/Modules/FindBISON.cmake:226 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
document/src/vespa/document/select/CMakeLists.txt:3 (find_package)

-- Configuring incomplete, errors occurred!
See also "/source/CMakeFiles/CMakeOutput.log".

**Comment by vekterli** [comment]
I'm assuming this is related to the build failure: https://github.com/vespa-engine/docker-image-dev/pull/10

Build deps were added to CentOS and internal building prior to merging, but docker was left hanging.

**Pull request merged**